src/Microsoft.ML.Data/Transforms/HashTransform.cs

```diff
@@ -119,52 +114,68 @@ public bool TryUnparse(StringBuilder sb)
+            /// <param name="hashBits">Number of bits to hash into. Must be between 1 and 31, inclusive.</param>
+            /// <param name="seed">Hashing seed.</param>
+            /// <param name="ordered">Whether the position of each term should be included in the hash.</param>
+            /// <param name="invertHash">Limit the number of keys used to generate the slot name to this many. 0 means no invert hashing, -1 means no limit.</param>
+            public ColumnInfo(string input, string output,
+                int hashBits = HashEstimator.Defaults.HashBits,
+                uint seed = HashEstimator.Defaults.Seed,
+                bool ordered = HashEstimator.Defaults.Ordered,
+                int invertHash = HashEstimator.Defaults.InvertHash)
             {
-                HashBits = col.HashBits ?? args.HashBits;
-                if (HashBits < NumBitsMin || HashBits >= NumBitsLim)
-                    throw Contracts.ExceptUserArg(nameof(args.HashBits), "Should be between {0} and {1} inclusive", NumBitsMin, NumBitsLim - 1);
-                HashSeed = col.Seed ?? args.Seed;
-                Ordered = col.Ordered ?? args.Ordered;
+                Input = input;
+                Output = output;
+                HashBits = hashBits;
+                Seed = seed;
+                Ordered = ordered;
+                InvertHash = invertHash;
             }
 
-            public ColInfoEx(ModelLoadContext ctx)
+            internal ColumnInfo(string input, string output, ModelLoadContext ctx)
             {
+                Input = input;
+                Output = output;
                 // *** Binary format ***
                 // int: HashBits
                 // uint: HashSeed
                 // byte: Ordered
-
                 HashBits = ctx.Reader.ReadInt32();
-                Contracts.CheckDecode(NumBitsMin <= HashBits && HashBits < NumBitsLim);
-
-                HashSeed = ctx.Reader.ReadUInt32();
+                Contracts.CheckDecode(HashEstimator.NumBitsMin <= HashBits && HashBits < HashEstimator.NumBitsLim);
+                Seed = ctx.Reader.ReadUInt32();
                 Ordered = ctx.Reader.ReadBoolByte();
             }
 
-            public void Save(ModelSaveContext ctx)
+            internal void Save(ModelSaveContext ctx)
             {
                 // *** Binary format ***
                 // int: HashBits
                 // uint: HashSeed
                 // byte: Ordered
 
-                Contracts.Assert(NumBitsMin <= HashBits && HashBits < NumBitsLim);
+                Contracts.Assert(HashEstimator.NumBitsMin <= HashBits && HashBits < HashEstimator.NumBitsLim);
                 ctx.Writer.Write(HashBits);
 
-                ctx.Writer.Write(HashSeed);
+                ctx.Writer.Write(Seed);
                 ctx.Writer.WriteBoolByte(Ordered);
             }
         }
 
-        private static string TestType(ColumnType type)
+        public static string TestType(ColumnType type)
```

Comment:
>public [](start = 8, length = 6)

why public? If it's public, it has to be documented

Reply:
Yea, I see now, you are using the same code in the estimator. I think you should turn it into `public static bool IsColumnTypeValid(ColumnType type)` and `internal const string ExpectedColumnType` to feed to the exceptions

---
In reply to: [218906546](https://github.com/dotnet/machinelearning/pull/944#discussion_r218906546) [](ancestors = 218906546)

Reply:
this signature is easier to document, and actually helpful to the consumer. Might as well also move it to the estimator.

---
In reply to: [218910693](https://github.com/dotnet/machinelearning/pull/944#discussion_r218910693) [](ancestors = 218910693,218906546)